shared/metadata.go

```diff
@@ -20,3 +50,94 @@ type MetadataLink struct {
 	TestPath string `yaml:"test"`
 	URL      string
 }
+
+func (m MetadataResults) Len() int           { return len(m) }
+func (m MetadataResults) Swap(i, j int)      { m[i], m[j] = m[j], m[i] }
+func (m MetadataResults) Less(i, j int) bool { return m[i].String() < m[j].String() }
+
+func (m MetadataResult) String() string {
+	var urlString string
+	for _, url := range m.URLs {
+		urlString += url
+	}
+	return m.Test + urlString
+}
+
+// GetMetadataResponse retrieves the response to a WPT Metadata query.
+func GetMetadataResponse(testRuns []TestRun, client *http.Client) MetadataResponse {
+	metadataByteMap, err := util.CollectMetadata(client)
+	if err != nil {
+		return MetadataResponse{}
+	}
+	metadata := parseMetadata(metadataByteMap)
+	return constructMetadataResponse(testRuns, metadata)
+}
+
+// parseMetadata collects and parses all META.yml files from
+// wpt-metadata reposiroty.
+func parseMetadata(metadataByteMap map[string][]byte) map[string]Metadata {
+	var metadataMap = make(map[string]Metadata)
+	for path, data := range metadataByteMap {
+		var metadata Metadata
+		err := yaml.Unmarshal(data, &metadata)
+		if err != nil {
+			log.Warningf("Failed to unmarshal %s.yml.", path)
+			continue
+		}
+		metadataMap[path] = metadata
+	}
+	return metadataMap
+}
+
+// ConstructMetadataResponse constructs the response to a WPT Metadata query.
+// When parsing 'link' nodes, assume there is no mising information nor duplicates;
+// assume each test for each browser type is only associated with one bug.
+func constructMetadataResponse(testRuns []TestRun, metadata map[string]Metadata) MetadataResponse {
+	res := MetadataResults{}
+	for folderPath, data := range metadata {
+		testMap := make(map[string][]string)
+
+		for _, link := range data.Links {
+			var urls []string
+
+			var fullTestName = ""
+			if strings.HasPrefix(link.TestPath, folderPath) {
+				fullTestName = link.TestPath
+			} else {
+				fullTestName = path.Join(folderPath, link.TestPath)
+			}
+
+			if _, ok := testMap[fullTestName]; !ok {
+				testMap[fullTestName] = make([]string, len(testRuns))
+			}
+			urls = testMap[fullTestName]
+
+			for i, run := range testRuns {
+				// Only matches browser type.
+				if link.Product.BrowserMatches(run) {
+					urls[i] = link.URL
+				}
+			}
+		}
+		for nameKey, urlsVal := range testMap {
+			isMatches := false
+
+			for _, url := range urlsVal {
+				if url != "" {
+					isMatches = true
+				}
+			}
+
+			// No matching testRuns.
+			if !isMatches {
```

Comment:
Once you've done the above, you can simply check whether `urlsVal` is nil here.

Reply:
relate to the comment above